content/courses/ada-in-practice/chapters/streams_device_driver_api_flexibility.rst

```diff
@@ -0,0 +1,662 @@
+of device independence, but did not do so for the sake of keeping the example
+simple.)
+
+Our solution will consist of an ADT named :ada:`Serial_Port`, and a means for
+sending and receiving values of higher-level types via :ada:`Serial_Port`
+objects. Type :ada:`Serial_Port` will be a wrapper for the device driver's
+USART type. Therefore, the type :ada:`Serial_Port` will have an access
+discriminant designating a USART:
+
+.. code-block:: ada
+
+       type Serial_Port (Device : not null access USART) is ...
+
+Using this :ada:`Device` discriminant a :ada:`Serial_Port` object can reference
+the wrapped physical USART in order to send and receive values via that
+hardware device.
+
+With that introduction in place, we can consider the possible approaches to
+sending and receiving values of higher-level types via :ada:`Serial_Port`
+objects.
+
+The canonical Ada approach consists of a generic package with a generic formal
+parameter type. That formal type represents a client-specific type to be sent
+or received. Clients instantiate the generic for every client-defined type
+necessary.
+
+The generic package would look like the following:
+
+.. code-block:: ada
+
+       generic
+          type Client_Data (<>) is limited private;
+       package Client_IO is
+
+          procedure Send
+            (This     : in out Serial_Port;
+             Outgoing : Client_Data);
+
+          procedure Receive
+            (This     : in out Serial_Port;
+             Incoming : out Client_Data);
+
+       end Client_IO;
+
+In the procedure bodies, the values of the :ada:`Incoming` or :ada:`Outgoing`
+parameters would be converted to or from bytes as necessary and sent or
+received via the USART designated by :ada:`This.Device`.
+
+This approach supports as many client-level types as required, including
+limited types. It is type-safe so the compiler can catch errors in the type(s)
+being sent and received. In addition, the low-level implementation details,
+such as unchecked conversions, are hidden inside the generic package body.
+
+Moreover, the approach is independent of other design considerations, such as
+whether callers wait for completion of the invoked I/O operation. The bodies of
+the generic procedures can be implemented to provide the expected behavior.
+
+However, this approach is somewhat heavy because the generic package must be
+instantiated for every client type to be supported. If there are many such
+types, there will be many instantiations. Not only is that more lines of source
+code, but also more object code because most Ada implementations do not support
+code-sharing for generic instantiations. But that said, the need for a given
+client to send and receive values of many different types is not typical.
+
+However, there is a more concise approach possible, for both the driver and
+client source code. This alternative approach leverages the flexibility of
+streams and stream attributes.
```

Comment:
At the end of this paragraph let's add this sentence:
Using streams allows the wrapper to support an unlimited number of distinct client types, with no additional source code required per type.